syslog: add a simple syslog module

vincent-d opened the pull request:
### Contribution description

This adds a simple syslog module that prints string whose format is compatible with rfc5424.

This module is organized into 2 layers:
- a core that receives logs, format them, and dispatch them to backends
- backends that actually print strings. At the moment stdio and file backend are implemented.

stdio backend simply prints out to stdio.
File backend prints in a file and take care of file rotation (configurable at build time).

API-wise, it is inspired by Linux' syslog API. The entry point is `(v)syslog` and optionally `openlog` and `closelog` can be built to allow adding different log entries.

### Testing procedure

`tests/syslog` provides some simple commands to do basic tests.

### Issues/PRs references

Some parts of header files were initially inspired by #11362
Closes #2376

maribu commented on sys/syslog/backend.h:
```diff
@@ -0,0 +1,109 @@
+struct syslog_msg {
+    /* metadata */
+    uint8_t pri;
+    struct tm time;
```
As above, IMO it would be better to use ztimer. A `ztimer_now_t` is also only 8 bytes in size (with `ztimer_now64` used), while `struct tm` is in between 36 and 44 Bytes on ARM. The backend could use a helper function to convert this to year, month, day, hour, minute and seconds.

It might be reasonable to add a `tm_t` type that contains only the reasonable fields using reasonable data times. IMO we should also consider updating the `periph_rtc` API to use that, so that driver authors don't need to worry about the day of the week, daylight saving settings, etc. in addition to saving same bytes for RAM. Maybe such a `tm_t` is better than `ztimer_now_t` for this use case.

vincent-d replied:
Unfortunately, there is no high-level API that takes advantage of RTC. The 'real' date is printed in the logs, so RTC is handy.
Though I agree, the full `struct tm` is a bit overkill and it may be too limiting to force the dependency to `periph_rtc`.
We could imagine having different implementations depending on if RTC is available I guess.

maribu commented on sys/syslog/backend.h:
```diff
@@ -0,0 +1,109 @@
+ * @brief   syslog backend driver
+ */
+typedef struct {
+    /**
+     * @brief   syslog backend init
+     */
+    int (*init)(void);
```
Can't we get rid of this function pointer and just initialize the backend within `auto_init` one by one?

vincent-d replied:
yes, I guess we could. I initially thought it would be handy to have this, but I don't see a lot of added value.

maribu commented on sys/syslog/backend_file.c:
```diff
@@ -0,0 +1,211 @@
+#include "vfs.h"
+#include "mutex.h"
+
+#define ENABLE_DEBUG    (1)
```
```suggestion
#define ENABLE_DEBUG    0
```

Please don't use braces here, as this prevents using it with the `IS_ACTIVE()` helper macro.

maribu reviewed (commented):
Some comments for the API. I didn't look at the implementation yet.

aabadie commented on core/include/log.h:
```diff
@@ -49,12 +52,14 @@ extern "C" {
-    LOG_ERROR,      /**< Error log level, will print only critical,
-                         non-recoverable errors like hardware initialization
-                         failures */
+    LOG_NONE = -1,
```
I think that changing this enum is a strong API change regarding the log level. In master, the debug log level corresponds to the value 4, with this change, it would become 7 if I'm not mistaken.

I can see that at least the `tests/log_color` test application won't work anymore as expected because of this:

https://github.com/RIOT-OS/RIOT/blob/2e2f950ca7ae0a455d48613c668dace842064a0c/tests/log_color/Makefile#L5-L6

There might be other applications affected as well.

vincent-d replied:
Would this really be an issue? I mean we can update the test apps (a proper way could be using `CFLAGS += -DLOG_LEVEL=LOG_DEBUG`). IMO any user code should use the enum and not the value directly...

aabadie replied:
When directly used with CFLAGS, the enum can indeed be used directly and it's clearer.
There's also the corresponding Make variable `LOG_LEVEL` and its documentation that should be adapted:

https://github.com/RIOT-OS/RIOT/blob/bd5b195433d42ca5f29d348f7b967e609e33452a/makefiles/vars.inc.mk#L119

miri64 replied:
Seems like the ESP32 SDK isn't too happy about this API change either :-/ https://ci.riot-os.org/RIOT-OS/RIOT/15739/c2c11e58d4921c710c5d776a686d3f239831829e/output/compile/tests/driver_apds99xx/esp32-ethernet-kit-v1_1:gnu.txt

vincent-d replied:
See my last commits for a workaround. 

If it's not ok with you, we can modify the ESP32 SDK, but I'm not sure what it exactly implies.

vincent-d commented on sys/include/syslog.h:
```diff
@@ -0,0 +1,276 @@
+
+/* STDIO */
+
+/**
+ * @brief   Enable/disable stdio output
+ *
+ * @param[in]   print   true to enable stdio output, false to disable
+ */
+void syslog_backend_stdio_print(bool print);
+
+/* File */
+
+/**
+ * @brief   Start file backend
+ *
+ * Start syslog file backend. This should be called when the filesystem is
+ * ready. It will initialize the file backend by looking into the log directory
+ * and counting the current number of files.
+ */
+int syslog_backend_file_start(void);
+
+/**
+ * @brief   Suspend log rotation
+ *
+ * If called multiple times, syslog_backend_file_resume_rotation() must be
+ * called the same number of times so the rotation is resumed.
+ */
+void syslog_backend_file_suspend_rotation(void);
+
+/**
+ * @brief   Resume log rotation
+ */
+void syslog_backend_file_resume_rotation(void);
```
I want to be able to upload files to a backend so that the easiest way of ensuring the file is not removed during the upload operation, while logging is not completely stopped.

vincent-d commented on sys/syslog/syslog.c:
```diff
@@ -0,0 +1,621 @@
+
+static void _set_time(struct syslog_msg *msg)
+{
+#if defined(MODULE_PERIPH_RTC)
+    rtc_get_time(&msg->time);
+    msg->time_set = true;
+#else
+    msg->time_set = false;
+#endif
```
@maribu RTC was actually already optional ;). So I guess we can get rid of the `struct tm` in the `struct syslog_msg` and use something more lightweight.

vincent-d commented:
@maribu I addressed some of your comments, and also split the "formatter" in a separate file, so one can implement a light header formatting easily.

jia200x commented:
are there chances to get this in by today? :)

miri64 commented:
> are there chances to get this in by today? :)

As there is an API change in here, that potentially breaks ESP support I would say this falls under soft-feature freeze.

vincent-d commented:
I reworked a bit the `formatter` part so it's easier to add a new formatter (quickly tried with a custom JSON formatter, but it would need more work to be usable).
I also added the option to add custom user-provided backends.

I think this is now ready ;)

fjmolinas commented on tests/syslog/Makefile:
```diff
@@ -0,0 +1,33 @@
+# To build openlog/closelog
+CFLAGS += -DCONFIG_SYSLOG_HAS_OPEN=1
+# Limit the number of entries to save RAM
+CFLAGS += -CONFIG_SYSLOG_MAX_ENTRIES=2
```
```suggestion
CFLAGS += -DCONFIG_SYSLOG_MAX_ENTRIES=2
```

fjmolinas commented:
Is this still wanted @vincent-d @maribu @aabadie?
